google-cloud-pubsub/src/main/java/com/google/cloud/pubsub/v1/MessageDispatcher.java

```diff
@@ -372,13 +400,54 @@ void processReceivedMessages(List<ReceivedMessage> messages) {
         // totally expire so that pubsub service sends us the message again.
         continue;
       }
-      outstandingBatch.add(new OutstandingMessage(message, ackHandler));
+      OutstandingMessage outstandingMessage = new OutstandingMessage(message, ackHandler);
+      if (this.exactlyOnceDeliveryEnabled.get()) {
+        outstandingReceipts.put(
+            message.getAckId(), new ReceiptCompleteData(outstandingMessage, false));
+      } else {
+        outstandingBatch.add(outstandingMessage);
+      }
       pendingReceipts.add(ackRequestData);
     }
-
     processBatch(outstandingBatch);
   }
 
+  synchronized void notifyAckSuccess(AckRequestData ackRequestData) {
+
+    if (outstandingReceipts.containsKey(ackRequestData.getAckId())) {
+      ReceiptCompleteData receiptCompleteData = outstandingReceipts.get(ackRequestData.getAckId());
+      OutstandingMessage outstandingMessage = receiptCompleteData.getOutstandingMessage();
+      // Setting to true means that the receipt is complete
+      receiptCompleteData.setReceiptComplete(true);
+      outstandingReceipts.put(ackRequestData.getAckId(), receiptCompleteData);
```

Comment:
You don't need to put the receiptCompleteData back into the outstandingReceipts, it was never removed.

Reply:
this is just updating the value of the ackRequestData.getAckId() to true

Reply:
Ah scratch that, get() returns a reference pointer back so we dont need to put it back. will change.

Reply:
have removed the put statement.